plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -15,89 +15,165 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Streams;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.trino.plugin.hive.util.AsyncQueue;
+import io.trino.plugin.hive.util.ResumableTask;
+import io.trino.plugin.hive.util.ResumableTasks;
+import io.trino.plugin.hive.util.ThrottledAsyncQueue;
 import io.trino.spi.TrinoException;
 import io.trino.spi.connector.ConnectorPartitionHandle;
 import io.trino.spi.connector.ConnectorSplit;
 import io.trino.spi.connector.ConnectorSplitSource;
 import io.trino.spi.connector.SchemaTableName;
 import org.apache.iceberg.CombinedScanTask;
 import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.TableScan;
 import org.apache.iceberg.io.CloseableIterable;
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
-import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Iterator;
-import java.util.List;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
 
-import static com.google.common.collect.Iterators.limit;
+import static io.airlift.concurrent.MoreFutures.addExceptionCallback;
+import static io.airlift.concurrent.MoreFutures.toCompletableFuture;
 import static io.trino.plugin.iceberg.IcebergUtil.getPartitionKeys;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
 import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
 import static java.util.Objects.requireNonNull;
-import static java.util.concurrent.CompletableFuture.completedFuture;
 
 public class IcebergSplitSource
         implements ConnectorSplitSource
 {
-    private final SchemaTableName schemaTableName;
-    private final CloseableIterable<CombinedScanTask> combinedScanIterable;
-    private final Iterator<FileScanTask> fileScanIterator;
+    private final AsyncQueue<ConnectorSplit> splitQueue;
+    private final IcebergSplitLoader loaderTask;
+    private final Executor executor;
+    private Throwable loaderException;
```

Comment:
It looks like there should be synchronization between thread which sets and reads loaderException.
Ot you have to make field `volatile`

Reply:
I would suggest to make field `AtomicReference` and pass instance of reference to callback.